### PR TITLE
[6.0] Fix the tiering delay to account for changing IL/native code versions during the delay

### PR DESCRIPTION
**Pull request by kouvel**
- Port of https://github.com/dotnet/runtime/pull/78668
- Added a check to see if the currently active native code version has a code entry point

Fixes https://github.com/dotnet/runtime/issues/77973

## Customer Impact

Apps that use profiler APIs to modify a method's IL may occasionally crash. If a new IL code version is added to a method while it's queued for call counting during the tiering delay, after the delay expires a call counting stub may be created that when called, would cause the app to crash. A workaround is to disable tiered compilation.

## Regression?

No

## Testing

Reproed the issue by inducing the timing with code changes and verified the fix. Scanned over the few other paths where call counting stubs would be created and verified that all of them guarantee a non-null code entry point for the method.

## Risk

Low

### CODE REVIEW

**Comment by carlossanlop** [comment]
@kouvel for the January servicing release, we only have a window of **one day** to merge backports (from Nov. 29th to Nov. 30th). For that reason, I'm asking all backport submitters to make sure the PR is ready for me to just hit the merge button on that day.

If this is ready, please ask @jeffschwMSFT if he approves the backport. If approved, then:
- Get a code review sign-off from an area owner (@davmason / @noahfalk).
- Add the `servicing-consider` label.
- Send an email to Tactics requesting approval.

**Review by jeffschwMSFT — APPROVED** [review]
approved.  we will take for consideration in 6.0.x

**Review by mangod9 — APPROVED** [review]

**Comment by kouvel** [comment]
cc @mangod9 @tommcdon

**Review by noahfalk — APPROVED** [review]

**Review by davmason — APPROVED** [review]

**Comment by carlossanlop** [comment]
Branding has been done. Milestone is 6.0.13. Approved by Tactics. Signed-off by area owners. CI is green. No OOB package authoring changes needed.
Ready to merge. :shipit:

**Pull request merged**
